Add support for graceful worker shutdown

nezihyigitbasi opened the pull request:
This is the initial version of the graceful worker shutdown functionality (tested with automated randomized query submissions + shutdown requests). \cc @martint @cberner @dain 

**Note:** In this version due to a network glitch a worker can go from "shutting down" state to the "failed" state and then back to the "active" state (as that node will be removed from the known node states when it gets marked as "failed" -- see `DiscoveryNodeManager`.`nodeStates` map). And in the "active" state the coordinator will submit tasks to that shutting down worker and it is possible that these tasks may be received after the worker registers the task completion listeners, eventually causing those tasks to fail ungracefully. One possible way to prevent this is to reject these tasks at the workers (in `TaskResource`.`createOrUpdateTask()` if a worker is in the "shutting down" state) and let the coordinator re-schedule them (not sure whether this is easy to implement with the current scheduling code).

nezihyigitbasi commented on presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java:
```diff
@@ -0,0 +1,129 @@
+
+            log.info("Waiting for all tasks to finish");
+
+            //TODO do we want a MAX_TASK_WAIT or want to wait until all tasks are done no matter how long they take?
```
Do we want a MAX_TASK_WAIT or want to wait until all tasks are done no matter how long they take?

cberner replied:
At least for us we don't need that, do you? If we want to set a timeout like that, we will just send a `kill -9` to the worker process when the timeout is up

nezihyigitbasi commented on presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java:
```diff
@@ -0,0 +1,129 @@
+            }
+
+            // wait for another grace period after all tasks complete
+            //TODO I am not sure whether we need this. Can there be any in-flight messages from the
```
I am not sure whether we need this. Can there be any in-flight messages from the coordinator to the worker after a task is done?

cberner replied:
Yes. The coordinator polls for the state of the task until it observes a "terminal" state, so we want this grace period to be long enough for the coordinator to observe the state

cberner commented on presto-main/src/main/java/com/facebook/presto/connector/system/NodeSystemTable.java:
```diff
@@ -44,7 +50,7 @@
             .column("node_version", VARCHAR)
             .column("coordinator", BOOLEAN)
-            .column("active", BOOLEAN)
+            .column("state", VARCHAR)
```
@dain thoughts on changing this? Should we make this backwards compatible and keep the old field?

martint replied:
This is mostly informational (for users), so I don't think it's a big issue to change it.

cberner replied:
:+1:

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java:
```diff
@@ -69,15 +86,56 @@
+
+                // Add new nodes
+                for (Node node : activeNodes) {
+                    if (!nodeStates.containsKey(node.getNodeIdentifier())) {
```
Use `putIfAbsent`?

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java:
```diff
@@ -165,16 +261,33 @@ public synchronized AllNodes getAllNodes()
+        switch (state) {
+            case ACTIVE:
+                return activeNodesByDataSource.get(datasourceName);
+            case INACTIVE:
```
Does anyone call this function with `INACTIVE` or `SHUTTING_DOWN`? Since it returns a constant, those calls probably don't make sense. I would switch this back to `getActiveDatasourceNodes` unless there's a reason we need to support this

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java:
```diff
@@ -105,7 +164,22 @@ private synchronized PrestoNode refreshNodesInternal()
             NodeVersion nodeVersion = getNodeVersion(service);
             if (uri != null && nodeVersion != null) {
-                PrestoNode node = new PrestoNode(service.getNodeId(), uri, nodeVersion);
+                boolean isActive = isActive(nodeVersion);
```
Instead of changing `isActive` to take a `NodeVersion` I would rename it to `getNodeState(PrestoNode node)` and move all the below logic for figuring out the state into that function

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/InMemoryNodeManager.java:
```diff
@@ -59,21 +60,38 @@ public void addNode(String datasourceName, Iterable<Node> nodes)
+        switch (state) {
+            case ACTIVE:
+                return ImmutableSet.copyOf(remoteNodes.get(datasourceName));
+            case INACTIVE:
```
Same comment as on the other node manager

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/RemoteNodeState.java:
```diff
@@ -0,0 +1,115 @@
+        return nodeState.get();
+    }
+
+    public void asyncRefresh()
```
I would make this `synchronized` just to be sure that multiple threads can't call it at the same time, and cause multiple requests to be sent

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/RemoteNodeState.java:
```diff
@@ -0,0 +1,115 @@
+            Futures.addCallback(responseFuture, new FutureCallback<JsonResponse<NodeState>>() {
+                @Override
+                public void onSuccess(@Nullable JsonResponse<NodeState> result)
+
```
extra newline

cberner commented on presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java:
```diff
@@ -139,6 +143,14 @@ public void configure(Binder binder)
         newExporter(binder).export(NodeScheduler.class).withGeneratedName();
 
+        // for polling worker states for graceful shutdown
+        httpClientBinder(binder).bindHttpClient("discoverNodeManager", ForGracefulShutdown.class)
```
typo. should be "discoveryNodeManager"

cberner commented on presto-main/src/main/java/com/facebook/presto/server/ServerInfoResource.java:
```diff
@@ -14,30 +14,68 @@
     }
+
+    @POST
+    @Path("shutdown")
```
We like to make all our endpoints RESTful, so this should be `@Path("state")` and should accept the new state to transition to

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java:
```diff
@@ -145,6 +226,21 @@ private synchronized PrestoNode refreshNodesInternal()
+        // check shutdownFlag for the current node's state
+        // for other nodes check nodeStates map (which is non-empty only on coordinator)
+        if (nodeId.equals(nodeInfo.getNodeId())) {
+            return shutdownFlag.get();
```
I don't think you need this `shutdownFlag`. The coordinator should be able to talk to itself to retrieve it's own state, and then it looks like you can remove the whole `ShutdownFlag` class

cberner commented:
Had some comments. Also, can you squash your commits together? I think the whole feature should be a single commit

nezihyigitbasi commented:
@cberner Addressed your comments & squashed the commits. How do we want to handle state transitions to ACTIVE and INACTIVE in `ServerInfoResource::updateState()`?

nezihyigitbasi commented:
As discussed, currently state transitions to ACTIVE and INACTIVE are not supported.

dain commented on presto-main/src/main/java/com/facebook/presto/server/ServerInfoResource.java:
```diff
@@ -14,30 +14,82 @@
+
+    @GET
+    @Path("state")
+    @Produces(APPLICATION_JSON)
```
Should this be `TEXT_PLAIN`

nezihyigitbasi replied:
yep we can make it plain text, I don't see any strong reason why it should be json.

nezihyigitbasi replied:
BTW I just tried making it plain text and got some weird error. So I am leaving it as is unless you really want it to me text/plain.

```
ERROR	http-worker-112	org.glassfish.jersey.message.internal.WriterInterceptorExecutor	MessageBodyWriter not found for media type=text/plain, type=class com.facebook.presto.spi.NodeState, genericType=class com.facebook.presto.spi.NodeState.
```

cberner replied:
I believe you need to return a `String` for "text/plain"

cberner replied:
Oh, I see why you have this as json. The `RemoteNodeState` takes JSON. I think that's fine

dain commented on presto-main/src/main/java/com/facebook/presto/server/ServerInfoResource.java:
```diff
@@ -14,30 +14,82 @@
+    @POST
+    @Path("state")
+    @Produces(TEXT_PLAIN)
+    public Response updateState(NodeState state)
```
Is this case sensitive?  If so, do we want it to be case sensitive?

dain replied:
Should this be idempotent?

nezihyigitbasi replied:
This is not case sensitive right now. For our use case I don't think it's a big deal for us, but if you think it's nice to have for you I can find a solution for that.

Isn't this operation already idempotent? Caller can call `updateState()`` with shutting_down multiple times and the net effect will be the same. Similarly calls with active/inactive are idempotent too as the net result is always failure.

cberner replied:
I think @dain meant that it shouldn't through an exception if it's already shutting down and you tell it to transition to `SHUTTING_DOWN` a second time

cberner commented on presto-main/src/main/java/com/facebook/presto/connector/system/SystemSplitManager.java:
```diff
@@ -74,7 +75,7 @@ public ConnectorSplitSource getSplits(ConnectorSession session, ConnectorTableLa
         }
         else if (tableDistributionMode == ALL_NODES) {
-            nodes.addAll(nodeManager.getActiveNodes());
+            nodes.addAll(nodeManager.getNodes(ACTIVE));
```
I think this also needs to include shutting down nodes. Otherwise you won't be able to query the tasks system table on those nodes

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java:
```diff
@@ -69,15 +84,53 @@
 
     @Inject
-    public DiscoveryNodeManager(@ServiceType("presto") ServiceSelector serviceSelector, NodeInfo nodeInfo, FailureDetector failureDetector, NodeVersion expectedNodeVersion)
+    public DiscoveryNodeManager(@ServiceType("presto") ServiceSelector serviceSelector, NodeInfo nodeInfo, FailureDetector failureDetector,
```
Put each argument on its own line

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java:
```diff
@@ -69,15 +84,53 @@
+    {
+        // poll worker states only on the coordinators
+        if (getCoordinators().contains(currentNode)) {
+            ScheduledExecutorService nodeStateUpdateExecutor = newSingleThreadScheduledExecutor(threadsNamed("node-state-poller-%s"));
```
Make this a field of the class, and shut it down in a `@PreDestroy` method

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java:
```diff
@@ -69,15 +84,53 @@
+            nodeStateUpdateExecutor.scheduleWithFixedDelay(() -> {
+                ImmutableSet.Builder nodeSetBuilder = ImmutableSet.builder();
+                AllNodes allNodes = getAllNodes();
+                Set<Node> activeNodes = nodeSetBuilder
```
name this variable "aliveNodes" since it's both active and shutting down nodes

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java:
```diff
@@ -69,15 +84,53 @@
+                        .addAll(allNodes.getShuttingDownNodes())
+                        .build();
+
+                ImmutableSet<String> activeNodeIds = activeNodes.stream()
```
`aliveNodeIds`

cberner commented on presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java:
```diff
@@ -139,6 +143,14 @@ public void configure(Binder binder)
         newExporter(binder).export(NodeScheduler.class).withGeneratedName();
 
+        // for polling worker states for graceful shutdown
+        httpClientBinder(binder).bindHttpClient("discoveryNodeManager", ForGracefulShutdown.class)
```
Name this "node-manager" instead of "discoveryNodeManager" since this name will show up in the configs

cberner commented on presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java:
```diff
@@ -0,0 +1,126 @@
+    private final AtomicBoolean shutdownRequested = new AtomicBoolean(false);
+
+    @Inject
+    public GracefulShutdownHandler(TaskManager sqlTaskManager, LifeCycleManager lifeCycleManager)
```
Inject `ServerConfig` as well, and add a check in `requestShutdown` that disallows shutting down a coordinator

cberner commented on presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java:
```diff
@@ -0,0 +1,126 @@
+            // a task may already be done when the listener is registered, but it's
+            // safe as the listener will be immediately called for a task in a terminal state
+            for (TaskInfo taskInfo : allTasks) {
+                log.info("Adding a state change listener to task %s", taskInfo);
```
Lets remove this logging, I think this is a little too verbose

cberner commented on presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java:
```diff
@@ -0,0 +1,126 @@
+    private final ExecutorService lifeCycleStopper = newSingleThreadExecutor(threadsNamed("lifecycle-stopper-%s"));
+    private final TaskManager sqlTaskManager;
+    private final LifeCycleManager lifeCycleManager;
+    private final AtomicBoolean shutdownRequested = new AtomicBoolean(false);
```
use a normal `boolean` with `@GuardedBy("this")` and synchronize access to it

cberner commented on presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java:
```diff
@@ -0,0 +1,126 @@
+    {
+        log.info("Shutdown requested");
+
+        shutdownRequested.set(true);
```
check that shutdownRequested is not already `true`

nezihyigitbasi replied:
@cberner Do we want to throw an exception if `shutdownRequested` is already true or just ignore the call silently?

cberner replied:
Just ignore it. I think this function should be idempotent

cberner commented on presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java:
```diff
@@ -0,0 +1,126 @@
+
+        //wait for a grace period to start the shutdown sequence
+        shutdownHandler.schedule(() -> {
+            List<TaskInfo> allTasks = sqlTaskManager.getAllTaskInfo();
```
I think you need to put all of this `countDownLatch` into a `while (there are any tasks that are not done)` loop, because new tasks might come in after you start this `CountDownLatch`

cberner commented on presto-main/src/main/java/com/facebook/presto/server/ServerInfoResource.java:
```diff
@@ -14,30 +14,82 @@
         return serverInfo;
     }
+
+    @POST
```
@dain pointed out that this should be `@PUT` instead of `@POST`

cberner commented on presto-main/src/test/java/com/facebook/presto/metadata/TestDiscoveryNodeManager.java:
```diff
@@ -48,10 +60,21 @@
+        {
+            @Nullable
+            @Override
+            public Response apply(Request input)
```
Use a lambda instead?

cberner commented on presto-main/src/test/java/com/facebook/presto/metadata/TestDiscoveryNodeManager.java:
```diff
@@ -48,10 +60,21 @@
+            @Override
+            public Response apply(Request input)
+            {
+                return new TestingResponse(HttpStatus.OK, ArrayListMultimap.create(), new byte[] {});
```
Shouldn't this return a `NodeState` in the data?

cberner commented:
I think it would be good to have an end-to-end test for this. Here's one suggestion for doing that:
1) Add a `ShutdownAction` interface with an `onShutdown` method. The default implementation would just call `System.exit(0)`. Then make `GracefulShutdownHandler` call this, instead of `System.exit(0)` directly.
2) Add a test similar to `TestMemoryManager` that submits a query or two, then initiates a shutdown on one of the `TestingPrestoServer`s. Modify `TestingPrestoServer` to bind a `TestShutdownAction` which stores a boolean, instead of calling `System.exit`, then you can check that the flag is set in your test

cberner commented:
Had some more comments, but overall this is looking good!

nezihyigitbasi commented:
@cberner addressed the comments & rebased, thanks. I extracted a new `GracefulShutdownModule` for simplifying the testing with a test shutdown action and added unit tests in `TestGracefulShutdown`.

cberner commented on presto-main/src/main/java/com/facebook/presto/server/DefaultShutdownAction.java:
```diff
@@ -0,0 +1,23 @@
+ */
+package com.facebook.presto.server;
+
+public class DefaultShutdownAction implements ShutdownAction
```
style note: `implements` should be on the next line

cberner commented on presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java:
```diff
@@ -0,0 +1,154 @@
+    private final ExecutorService lifeCycleStopper = newSingleThreadExecutor(threadsNamed("lifecycle-stopper-%s"));
+    private final LifeCycleManager lifeCycleManager;
+    private final TaskManager sqlTaskManager;
+    private final ServerConfig serverConfig;
```
Because `Config` objects are mutable we try to avoid retaining references to them. Instead, make this a boolean field and copy the `isCoordinator` value out in the constructor

cberner commented on presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java:
```diff
@@ -0,0 +1,154 @@
+        this.lifeCycleManager = requireNonNull(lifeCycleManager, "lifeCycleManager is null");
+    }
+
+    public void requestShutdown()
```
Make this `synchronized`, otherwise there's a race where two threads could call this at the same time, both get past the `if` and then both schedule shutdown sequence tasks

nezihyigitbasi commented:
@cberner updated.

cberner commented on presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java:
```diff
@@ -90,6 +98,56 @@
+    private final ShutdownAction testShutdownAction;
+    private final boolean coordinator;
+
+    public static class TestShutdownAction implements ShutdownAction
```
style note: `implements` should be on next line

cberner commented on presto-main/src/main/java/com/facebook/presto/server/GracefulShutdownHandler.java:
```diff
@@ -0,0 +1,154 @@
+public class GracefulShutdownHandler
+{
+    private static final Logger log = Logger.get(GracefulShutdownHandler.class);
+    private static final Duration GRACE_PERIOD = new Duration(2, MINUTES);
```
Lets make this configurable, so that it can be set to a lower value in the test

nezihyigitbasi replied:
How about `shutdown.grace-period` in `ServerConfig`? Any other ideas?

cberner replied:
Ya, that sounds good to me

cberner commented on presto-tests/src/test/java/com/facebook/presto/tests/TestGracefulShutdown.java:
```diff
@@ -0,0 +1,90 @@
+    public void testShutdown()
+            throws Exception
+    {
+        ImmutableMap<String, String> properties = ImmutableMap.of("node-scheduler.include-coordinator", "false");
```
Reduce the grace period, so that this test doesn't take too long (I'd set it to 10 seconds)

cberner commented on presto-tests/src/test/java/com/facebook/presto/tests/TestGracefulShutdown.java:
```diff
@@ -0,0 +1,90 @@
+            }
+
+            //let the worker receive some tasks
+            Thread.sleep(5_000);
```
Remove this, and instead get the `TaskManager` from the `TestingPrestoServer` and wait until it has received tasks. You can just have a loop that spins with a short sleep until the tasks show up

cberner commented on presto-tests/src/test/java/com/facebook/presto/tests/TestGracefulShutdown.java:
```diff
@@ -0,0 +1,90 @@
+@Test(singleThreaded = true)
+public class TestGracefulShutdown
+{
+    private static final long SHUTDOWN_TIMEOUT_MILLIS = 600_000;
```
This is too long, lets make it 240_000. No tests should take this long to run

cberner commented on presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java:
```diff
@@ -90,6 +98,56 @@
+
+    public static class TestShutdownAction implements ShutdownAction
+    {
+        private Lock lock = new ReentrantLock();
```
I would replace the lock and condition with a `CountDownLatch` (initialize it to 1). Then in `onShutdown` call `countDown()` and in `waitForShutdownComplete` just wait on the latch

cberner commented on presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java:
```diff
@@ -90,6 +98,56 @@
+        private Lock lock = new ReentrantLock();
+        private Condition shutdownCalled = lock.newCondition();
+
+        @GuardedBy("lock")
```
You can just make this `@GuardedBy("this")` if you make the change above to use a `CountDownLatch`

cberner commented on presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java:
```diff
@@ -278,6 +349,21 @@ public ClusterMemoryManager getClusterMemoryManager()
+        return gracefulShutdownHandler;
+    }
+
+    public ShutdownAction getTestShutdownAction()
```
name this `getShutdownAction`

cberner commented on presto-main/src/main/java/com/facebook/presto/server/ServerInfoResource.java:
```diff
@@ -14,30 +14,80 @@
+    @PUT
+    @Path("state")
+    @Produces(TEXT_PLAIN)
+    public Response updateState(NodeState state)
```
Add `@Consumes(APPLICATION_JSON)` since it looks like this only accepts json

cberner commented on presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java:
```diff
@@ -90,6 +97,42 @@
+            return isWorkerShutdown;
+        }
+
+        public synchronized void setWorkerShutdown(boolean isWorkerShutdown)
```
Since this isn't used outside this class, I would inline it and make `onShutdown` synchronized instead

cberner commented on presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java:
```diff
@@ -90,6 +97,42 @@
+    public static class TestShutdownAction
+            implements ShutdownAction
+    {
+        private CountDownLatch shutdownCalled = new CountDownLatch(1);
```
Make this `final`

cberner commented on presto-tests/src/test/java/com/facebook/presto/tests/TestGracefulShutdown.java:
```diff
@@ -0,0 +1,99 @@
+
+            // wait until tasks show up on the worker
+            while (taskManager.getAllTaskInfo().size() == 0) {
+                Thread.sleep(500);
```
Use `MILLISECONDS.sleep(500)`

cberner commented on presto-tests/src/test/java/com/facebook/presto/tests/TestGracefulShutdown.java:
```diff
@@ -0,0 +1,99 @@
+            TaskManager taskManager = worker.getTaskManager();
+
+            // wait until tasks show up on the worker
+            while (taskManager.getAllTaskInfo().size() == 0) {
```
Use `isEmpty()` instead of `size() == 0`

cberner commented:
You've missing a test in `TestServerConfig` (see Travis)

cberner commented:
Looks good, except for the minor comments and the Travis failure

nezihyigitbasi commented:
@cberner fixed the travis failure & addressed the comments.

cberner commented:
Merged. Thanks!
